lib/pacemaker/pcmk_sched_instances.c

```diff
@@ -1133,40 +1133,69 @@ pcmk__instance_update_ordered_actions(pe_action_t *first, pe_action_t *then,
-
-    for (gIter = children; gIter != NULL; gIter = gIter->next) {
-        pe_action_t *child_action = NULL;
-        pe_resource_t *child = (pe_resource_t *) gIter->data;
-
-        child_action = find_first_action(child->actions, NULL, task_s, child->children ? NULL : node);
-        pe_rsc_trace(action->rsc, "Checking for %s in %s on %s (%s)", task_s, child->id,
-                     pe__node_name(node), child_action?child_action->uuid:"NA");
-        if (child_action) {
-            enum pe_action_flags child_flags = child->cmds->action_flags(child_action, node);
-
-            if (pcmk_is_set(flags, pe_action_optional)
-                && !pcmk_is_set(child_flags, pe_action_optional)) {
-                pe_rsc_trace(child, "%s is mandatory because of %s", action->uuid,
-                             child_action->uuid);
-                pe__clear_action_summary_flags(flags, action, pe_action_optional);
-                pe__clear_action_flags(action, pe_action_optional);
-            }
-            if (pcmk_is_set(child_flags, pe_action_runnable)) {
-                any_runnable = true;
-            }
+    enum pe_action_flags flags;
+    const char *action_name = orig_action_name(action);
+
+    // Set original assumptions (optional and runnable may be cleared below)
+    flags = pe_action_optional|pe_action_runnable|pe_action_pseudo;
+
+    for (const GList *iter = instances; iter != NULL; iter = iter->next) {
+        const pe_resource_t *instance = iter->data;
+        const pe_node_t *instance_node = NULL;
+        pe_action_t *instance_action = NULL;
+        enum pe_action_flags instance_flags;
+
+        // Node is relevant only to primitive instances
+        if (instance->variant == pe_native) {
+            instance_node = node;
+        }
+
+        instance_action = find_first_action(instance->actions, NULL,
+                                            action_name, instance_node);
+        pe_rsc_trace(action->rsc, "%s has %s for %s on %s",
+                     instance->id,
+                     (instance_action == NULL)? "no action" : instance_action->uuid,
+                     action_name, pe__node_name(node));
+        if (instance_action == NULL) {
+            continue;
+        }
+
+        instance_flags = instance->cmds->action_flags(instance_action, node);
+
+        // If any instance action is mandatory, so is the collective action
+        if (pcmk_is_set(flags, pe_action_optional)
+            && !pcmk_is_set(instance_flags, pe_action_optional)) {
+            pe_rsc_trace(instance, "%s is mandatory because %s is",
+                         action->uuid, instance_action->uuid);
+            pe__clear_action_summary_flags(flags, action, pe_action_optional);
+            pe__clear_action_flags(action, pe_action_optional);
+        }
+
+        // If any instance action is runnable, so is the collective action
+        if (pcmk_is_set(instance_flags, pe_action_runnable)) {
+            any_runnable = true;
```

Comment:
There is an opportunity to break out of the loop if we've already cleared the optional flag and set `any_runnable = true` (as long as we don't rely on any side effects from `instance->cmds->action_flags()`. It doesn't save much computation though, so I'm indifferent.

Reply:
Unfortunately it can have side effects